### DISCUSSION TITLE
darwin.apple_sdk.sdk (Intel build fails) and darwin.apple_sdk.MacOSX-SDK (arm build success).

### DISCUSSION THREAD

**Discussion by PlumpMath** [ipetkov/crane#329] [General]
Hello

help me~

I'm building the basic template of the Fyrox game engine. 

As the title suggests, while the build is successful on ARM Mac, it's not the case for Intel Mac. The code is ready to use with hidden buildInputs, but they are just the ones under the frameworks and a few like stdenv.cc.cc.lib, libiconv, lld, clang, and llvm. On Linux, it requires things like fontconfig and alsa-lib. The important thing here is the Mac build, and the Mac build requires COREAUDIO_SDK_PATH. It seems like darwin.apple_sdk.MacOSX-SDK is exclusively for ARM M1 and M2. Do you happen to know what the macOS SDK for Intel is? I want to successfully complete the build.

```
  basic3d-crane-edit = craneLib.buildPackage rec {
    pname = "basic3d-edit";
    version = "0.1.0";
    src = cleanCargoSource (craneLib.path ./.);
    cargoExtraArgs = "-p editor";
    cargoVendorDir = vendorCargoDeps { cargoLock = ./Cargo.lock; src = ./.; };
    nativeBuildInputs = buildInputs;
    buildInputs = buildInputs-full;
    COREAUDIO_SDK_PATH = if sinfo.distro == "mac-intel" then "${darwin.apple_sdk.sdk}" else "${MacOSX-SDK}";
  };

```

Thanks.

**Comment by figsoda** [comment]
`COREAUDIO_SDK_PATH` seems to be introduced by the `coreaudio-sys` dependency in its build script: https://github.com/RustAudio/coreaudio-sys/blob/9358a98f4968a3017e0471bd828a404fda4ef501/build.rs#L6, and it falls back to using `xcrun` to determine the path.

You might be able to get away with not setting `COREAUDIO_SDK_PATH` and just by adding `darwin.apple_sdk.frameworks.CoreAudio` to your `buildInputs`. There are also a few [examples](https://github.com/search?q=COREAUDIO_SDK_PATH+language%3ANix&type=code) of using `COREAUDIO_SDK_PATH` on GitHub, you can also take inspiration from those if just adding the framework to `buildInputs` doesn't work.

**Comment by PlumpMath** [comment]
After trying to update nixpkgs flake, I'm getting occur errors in m2! Now it doesn't recognize the MacOSX-SDK that used to work fine on my M2. I wonder what's going on all of a sudden. Haha. -_- Onz.
It seems that nixpkgs itself is being modified currently. I'm eagerly waiting for it to work, and I did check the link you recommended, but it seems to be an old version... 

https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/apple-sdk/default.nix
version = "10.12";

I feel like it would be a hassle to match the frameworks and versions in the current nixpkgs and constantly monitor them, so I'm just going to wait until nixpkgs provides a simple solution.
However, I couldn't wait any longer, so I'll probably give it a try within the next few days following the link you provided
Thanks.

**Comment by PlumpMath** [comment]
There haven't been any changes to nixpkgs darwin. It was my misunderstanding. The reason was that there were internal updates happening on my Mac, and when I installed the command-line tools again, it started working on my M2. Even referring to the link you provided for intel-mac, which includes darwin.apple-sdk.sdk, darwin.apple-sdk_11.sdk, and darwin.apple-sdk_11.MacOSX-SDK (the three default Intel-based options provided by nixpkgs), they all fail as well.

It's a bit disheartening. Perhaps nixpkgs has given up on supporting Intel Macs since the release of M1? After all, Apple is a closed-source company. Haha, I hope it will work someday.

**Comment by PlumpMath** [comment]
https://github.com/NixOS/nixpkgs/pull/176661

https://github.com/reckenrode/nixpkgs/blob/master/doc/stdenv/platform-notes.chapter.md

**Comment by reckenrode** [comment]
I was able to reproduce the issue installing `fyrox-template` and creating a new project named “basic3d”. I then copied the package definition from the OP and tried to build it. It did fail, but I was able to get it to build using the default SDKs in nixpkgs. See https://gist.github.com/reckenrode/44d1bb807effac7e453d02040483a53b for the `flake.nix` I used.

The key was adding `rustPlatform.bindgenHook` to the `nativeBuildInputs`. After I did that, coreaudio-sys was able to find the dependencies it needs, and I was able to build the package and run the editor successfully.

**Reply by figsoda** [reply]
was that aarch64-darwin or x86_64-darwin, because I ran into some rather undebuggable errors when testing the latter

**Reply by reckenrode** [reply]
That was aarch64-darwin. I kicked off an x86_64-darwin build after I replied. It eventually failed due to a missing symbol `_kCFURLVolumeAvailableCapacityForImportantUsageKey`, which is only in the 11.0 SDK or newer. My apologies for the bad assumption. I’ll update the flake in the gist and report back.

**Reply by reckenrode** [reply]
The report isn’t good. I’m getting presumably the same errors, which usually happens due to mixing SDKs.

Unfortunately, I’m not having any luck using the stdenv from the 11.0 SDK. I’ve updated the gist with what I have in case that helps someone figure out a solution.

**Reply by PlumpMath** [reply]
I also encounter the same error only on x86-64. The error with '_kCFURLVolumeAvailableCapacityForImportantUsageKey' is resolved by reinstalling the command line tools, but even after reinstalling them on x86-64, it still fails. Thank you for reproducing the issue. I hope someone from the NixOS community can provide a solution. 

As you pointed out, I don't believe that a different higher version (sdk 13), as you mentioned, will solve the issue of mixed versions. The versions available within nixpkgs are sdk 10.04? and 11.
Thank you.

https://github.com/RustAudio/coreaudio-sys/issues/78

https://github.com/FyroxEngine/Fyrox/issues/455

**Reply by reckenrode** [reply]
The coreaudio-sys issue is resolved by adding `rustPlatform.bindgenHook`, which is required for any crate that uses `bindgen`.  See the section in the documentation on [hooks][1] for Rust. However, that’s not the issue now. The issue now is the default SDK used by nixpkgs on x86_64-darwin is too old to build Fyrox. Something requires 11.0, but the SDK is 10.12. It is possible to use the newer one, but the usual pattern doesn’t compose very well with language frameworks, so one has to do it manually.

The problem here is headers from the 10.12 SDK are being pulled into the build from some package. I tried manually replacing the CF in the stdenv (which is probably a bug because that should be handled automatically), but it’s also pulling in Libsystem from the 10.12 SDK. Manually editing and sourcing the `env-vars` from the failed build allows it to succeed, so that must be the issue. The problem is figuring out which package in the stack is pulling in the unwanted headers, so one can override it.

**Reply by figsoda** [reply]
there is `darwin.apple_sdk_11_0`, which I tried and ran into a bunch of syntax errors for some reason, and replacing stdenv with a newer version of clang didn't help either

**Reply by reckenrode** [reply]
I updated the flake with the above changes and also manually overrode Rust’s stdenv with the updated one. It’s going to require rebuilding Rust unfortunately, which will take a while. I won’t be able to check on it until much later today to see if everything built successfully.

**Reply by figsoda** [reply]
I don't think overriding rust's stdenv would do anything, I did `rustPlatform.buildRustPackage.override { stdenv = clang16Stdenv; }`, which didn't need a rebuild of rust

**Reply by reckenrode** [reply]
> there is `darwin.apple_sdk_11_0`, which I tried and ran into a bunch of syntax errors for some reason, and replacing stdenv with a newer version of clang didn't help either

The standard pattern is to use `darwin.apple_sdk_11_0.callPackages`, which takes care of doing the necessary substitutions. Unfortunately, for language frameworks currently, that doesn’t work. You have to override and substitute everything manually.

The problem here is something needs to have its stdenv overridden, but it’s not, so the wrong Libsystem is also being pulled into the build. I have tried overriding the stdenv used to build Rust, but that means Rust will need to be rebuilt. That takes a while, so I don’t know if it worked yet. The updated flake is available though in the gist if someone else wants to try.

**Reply by figsoda** [reply]
This is what I have so far, it does seems to be using the correct `libSystem`, but I really don't know how to debug this error

`flake.nix`
```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
  };

  outputs = { self, nixpkgs }:
    let
      inherit (nixpkgs.lib)
        genAttrs
        ;
      eachSystem = f: genAttrs
        [
          "aarch64-darwin"
          "x86_64-darwin"
        ]
        (system: f nixpkgs.legacyPackages.${system});
    in
    {
      packages = eachSystem (pkgs: {
        default = pkgs.darwin.apple_sdk_11_0.callPackage self { };
      });
    };
}
```
`default.nix`
```nix
{ rustPlatform, clang16Stdenv, cmake, darwin }:

let
  inherit (darwin.apple_sdk_11_0.frameworks)
    AudioUnit
    CoreAudioKit
    OpenAL
    ;
in

rustPlatform.buildRustPackage.override { stdenv = clang16Stdenv; } {
  pname = "editor";
  version = "0.1.0";

  src = ./.;

  cargoLock = {
    lockFile = ./Cargo.lock;
  };

  nativeBuildInputs = [
    cmake
    rustPlatform.bindgenHook
  ];

  buildInputs = [
    AudioUnit
    CoreAudioKit
    OpenAL
  ];

  cargoBuildFlags = [ "-p" "editor" ];

  doCheck = false;
}
```
log
```
error: failed to run custom build command for `coreaudio-sys v0.2.12`

Caused by:
  process didn't exit successfully: `/private/tmp/nix-build-editor-0.1.0.drv-0/x3ksjslssscrrfgr3pi87p2g4palc3fj-source/target/release/build/coreaudio-sys-be2f37f882c61799/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=COREAUDIO_SDK_PATH
  cargo:rustc-link-lib=framework=AudioUnit
  cargo:rustc-link-lib=framework=AudioToolbox
  cargo:rustc-link-lib=framework=CoreAudio
  cargo:rustc-link-lib=framework=OpenAL
  cargo:rustc-link-lib=framework=CoreMIDI
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS

  --- stderr
  /nix/store/9dw29sl9gsl8090cf6qkmhj2pprl3z5z-apple-framework-CoreFoundation-11.0.0/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:1189:83: error: expected ','
  /nix/store/9dw29sl9gsl8090cf6qkmhj2pprl3z5z-apple-framework-CoreFoundation-11.0.0/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:1190:98: error: expected ','
  /nix/store/9dw29sl9gsl8090cf6qkmhj2pprl3z5z-apple-framework-CoreFoundation-11.0.0/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:1200:85: error: expected ','
  /nix/store/10s4dxrgp9gkcm6z10sn4rc423icbacb-apple-framework-AudioToolbox-11.0.0/Library/Frameworks/AudioToolbox.framework/Headers/AUComponent.h:1595:72: error: expected ','
  /nix/store/10s4dxrgp9gkcm6z10sn4rc423icbacb-apple-framework-AudioToolbox-11.0.0/Library/Frameworks/AudioToolbox.framework/Headers/AUComponent.h:1629:72: error: expected ','
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:48:1: error: expected ')'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:48:1: note: to match this '('
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:48:19: error: a parameter list without types is only allowed in a function definition
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:93:24: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:116:1: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:116:16: error: nullability specifier '_Nullable' cannot be applied to non-pointer type 'int'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:137:1: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:138:65: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:137:16: error: nullability specifier '_Nullable' cannot be applied to non-pointer type 'int'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:177:19: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:202:20: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:241:32: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:271:32: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:299:21: error: unknown type name 'os_workgroup_t'
  /nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:311:25: error: unknown type name 'os_workgroup_t'
  fatal error: too many errors emitted, stopping now [-ferror-limit=]
  thread 'main' panicked at 'unable to generate bindings: ClangDiagnostic("/nix/store/9dw29sl9gsl8090cf6qkmhj2pprl3z5z-apple-framework-CoreFoundation-11.0.0/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:1189:83: error: expected ','\n/nix/store/9dw29sl9gsl8090cf6qkmhj2pprl3z5z-apple-framework-CoreFoundation-11.0.0/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:1190:98: error: expected ','\n/nix/store/9dw29sl9gsl8090cf6qkmhj2pprl3z5z-apple-framework-CoreFoundation-11.0.0/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:1200:85: error: expected ','\n/nix/store/10s4dxrgp9gkcm6z10sn4rc423icbacb-apple-framework-AudioToolbox-11.0.0/Library/Frameworks/AudioToolbox.framework/Headers/AUComponent.h:1595:72: error: expected ','\n/nix/store/10s4dxrgp9gkcm6z10sn4rc423icbacb-apple-framework-AudioToolbox-11.0.0/Library/Frameworks/AudioToolbox.framework/Headers/AUComponent.h:1629:72: error: expected ','\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:48:1: error: expected ')'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:48:19: error: a parameter list without types is only allowed in a function definition\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:93:24: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:116:1: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:116:16: error: nullability specifier '_Nullable' cannot be applied to non-pointer type 'int'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:137:1: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:138:65: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:137:16: error: nullability specifier '_Nullable' cannot be applied to non-pointer type 'int'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:177:19: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:202:20: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:241:32: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:271:32: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:299:21: error: unknown type name 'os_workgroup_t'\n/nix/store/83pzmprwijh9n58ymwwnk367vpdcwfij-libSystem-11.0.0/include/os/workgroup_object.h:311:25: error: unknown type name 'os_workgroup_t'\nfatal error: too many errors emitted, stopping now [-ferror-limit=]\n")', /private/tmp/nix-build-editor-0.1.0.drv-0/cargo-vendor-dir/coreaudio-sys-0.2.12/build.rs:150:39
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

**Reply by reckenrode** [reply]
If you build with `--keep-failed` and check `env-vars`, what does it have? What I see is `-idirafter /nix/store/c834w8pg8ypl9lcaxqgj2bpsxbv518rn-Libsystem-1238.60.2/include `, which is being pulled in from _somewhere_. I just don’t know which dependency is doing it, so I can override its stdenv.

**Reply by reckenrode** [reply]
The problem is `bindgenHook`. It’s still using the 10.12 SDK headers even though we want to use a 11.0 SDK stdenv. The hook is pulling in `nix-support/libc-cflags` from the old clang instead of the new one. That’s the source of the above include that’s being added to `BINDGEN_EXTRA_CLANG_ARGS`. This is probably a bug in nixpkgs since the 11.0 SDK `rustPlatform` shouldn’t be pulling in any 10.12 SDK headers.

**Reply by figsoda** [reply]
it seems to be pulled in from cmake
```console
$ nix why-depends . /nix/store/c834w8pg8ypl9lcaxqgj2bpsxbv518rn-Libsystem-1238.60.2 --system x86_64-darwin --derivation
/nix/store/d9vn7p2krpl3abm8cfrd2k85f570mn09-editor-0.1.0.drv
└───/nix/store/7xakn627gwyjh459k0wdbnl2rmv7saj4-cmake-3.25.3.drv
    └───/nix/store/0ig7a5lm4kjgy894jjd3s2d10500fn9a-Libsystem-1238.60.2.drv
```

**Reply by reckenrode** [reply]
Dependencies can use the 10.12 SDK as long as they don’t propagate those build inputs, so that’s probably okay.

**Comment by figsoda** [comment]
thanks for the hint @reckenrode, adding libsystem to `buildInputs` fixes the issue for me. here are the nix files for the default template

`default.nix`
```nix
{ rustPlatform, cmake, darwin }:

let
  inherit (darwin.apple_sdk_11_0.frameworks)
    AudioUnit
    CoreAudioKit
    OpenAL
    ;
in

rustPlatform.buildRustPackage {
  pname = "editor";
  version = "0.1.0";

  srcs = ./.;

  cargoLock = {
    lockFile = ./Cargo.lock;
  };

  nativeBuildInputs = [
    cmake
    rustPlatform.bindgenHook
  ];

  buildInputs = [
    AudioUnit
    CoreAudioKit
    OpenAL
    darwin.apple_sdk_11_0.Libsystem
  ];

  cargoBuildFlags = [ "-p" "editor" ];
  cargoTestFlags = [ "-p" "editor" ];
}
```
`flake.nix`
```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
  };

  outputs = { self, nixpkgs }:
    let
      inherit (nixpkgs.lib)
        genAttrs
        ;
      eachSystem = f: genAttrs
        [
          "aarch64-darwin"
          "x86_64-darwin"
        ]
        (system: f nixpkgs.legacyPackages.${system});
    in
    {
      packages = eachSystem (pkgs: {
        default = pkgs.darwin.apple_sdk_11_0.callPackage self { };
      });
    };
}
```

**Reply by reckenrode** [reply]
It probably shouldn’t be necessary to manually specify Libsystem. That would be like having to provide glibc explictly on Linux.

Looking at the way bindgenHook is written, it’s using the `libc-cflags` from `clang`, which is still the 10.12 SDK clang. It should be using the 11.0 SDK one. I’ll open an issue in nixpkgs to get that fixed.

**Reply by PlumpMath** [reply]
It's going well.
Thank you both for your care and assistance. I love both of you and all the people involved. Health, health. Blessings and grace.
Wa~~~~~

**Reply by reckenrode** [reply]
I have a fix for the `bindgenHook` issue. It’s in my stdenv bump branch though, so it’s requiring a lot of stuff to be built. Once I confirm it works with this test case, I’ll open a PR. That should allow `rustPlatform.bindgenHook` to work properly with the 11.0 SDK on x86_64-darwin and fix the syntax errors that are happening when it tries to build coreaudio-sys.

**Reply by reckenrode** [reply]
The PR fixing the `rustPlatform.bindgenHook` issue is https://github.com/NixOS/nixpkgs/pull/233815.
